[HW5] (Application architecture)

AlwaysSayNo opened the pull request:
- Added general and extended diagrams.
- General diagram shows the general concepts of modules and their communication.
- Extended diagram shows communication among specific modules and more detailed infrastructure of application.

Commit:
[Test] (Added build step)

Commit:
[Refactoring] (Refactoring after PR)

Commit:
[Refactoring] (Refactored linet workflow)

andiserg commented on cmd/main.go:
```diff
@@ -49,6 +61,28 @@ func main() {
+	}
+}
+
+func getCurrencyProviderChain() servcurrency.Provider {
+	// GET PROVIDERS
+	privateClient, err := private.NewClient(config.LoadCurrencyServiceConfig())
+	if err != nil {
+		log.Fatal("creating Private Bank currency provider")
 	}
+	govUaClient, err := govua.NewClient(config.LoadCurrencyServiceConfig())
+	if err != nil {
+		log.Fatal("creating Bank Gov Ua currency provider")
+	}
+	jsDelivrClient, err := cdnjsdelivr.NewClient(config.LoadCurrencyServiceConfig())
+	if err != nil {
+		log.Fatal("creating JS Deliver currency provider")
+	}
+
+	// SET PROVIDERS CHAIN
+	govUaClient.SetNext(jsDelivrClient)
+	privateClient.SetNext(govUaClient)
+
+	return privateClient
```
what does this function do in cmd/main.go?

AlwaysSayNo replied:
This function setups raters chain

andiserg replied:
In my opinion, it should be placed in the currency module

andiserg commented on internal/service/user/service.go:
```diff
@@ -0,0 +1,61 @@
+		return nil, fmt.Errorf("saving savedUser in database: %w", err)
+	}
+
+	userDTO := dto.ToDTO(*savedUser)
```
It looks strange. A Data Transfer Object is a simple data structure that does not require knowledge of how to map other structures to it

AlwaysSayNo replied:
dto in this example is a package, not an object

Commit:
[Architecture] (Added architecture diagrams)

- Added general and extended diagrams.
- General diagram shows the general concepts of modules and their communication.
- Extended diagram shows communication among specific modules and more detailed infrastructure of application.

Commit:
[Rater] (Added rater mock)

Commit:
[Architecture tests] (Wrote architecture tests)

- Wrote architecture tests using arch-go library.

Commit:
[Workflow] (Wrote GitHub workflow)

- Wrote GitHub workflow for Arch-Go.

Commit:
[Workflow] (Actions refactoring)

- Refactored actions so they are called only once per push

AlwaysSayNo committed:
Merge branch 'main' into hw5-application-architecture
